### PR TITLE
Add external contract calling functionality

### PR DESCRIPTION
**Pull request by DavidKnott**
### - What I did
I added the ability to call external contracts in response to #297.
### - How I did it
I did this by allowing existing contracts to be referenced using:

        class insert_contract_name():
            _insert_existing_external_functions_you_need_to_call_here
To understand the formatting further look in my `tests/test_parser.py`.  @fubuloubu @JackieXu  I got this functionality working (it still has some hiccups).  When you have a second would you mind taking a look?
### - How to verify it
I added tests in `tests/test_parser.py`
### - Description for the changelog
Add external contract calling functionality
### - Cute Animal Picture
image

### CODE REVIEW

File: tests/test_parser.py [hunk 1]
```diff
@@ -495,7 +495,7 @@ def get_comb() -> num:
     return self.comb
     """
     c = get_contract_with_gas_estimation(constructor_advanced_code2, args=[[5,7], "dog", 8])
-    assert c.get_comb() == 5738
+    assert c.get_comb() == 5708
```
Comment (fubuloubu) [inline tests/test_parser.py]: Ugh, this is totally unhelpful without a comment. Why the change?

File: tests/test_parser.py [hunk 2]
```diff
@@ -1390,6 +1390,115 @@ def returnten() -> num:
...
+
+best_number: public(num)
+
+def __init__(arg1: address):
+    self.best_number = Bar(arg1).get_magic_number()
+    """
+
+    c3 = get_contract(contract_3, args=[c2.address])
+    assert c3.get_best_number() == lucky_number
+    print('Successfully executed a multiple external contract calls')
```
Comment (fubuloubu) [inline tests/test_parser.py]: I think you should include some cases where there are at least two external contracts referenced in your main testing contract, maybe setting the value in another from the value in one

File: tests/test_parser.py [hunk 3]
```diff
@@ -1390,6 +1390,115 @@ def returnten() -> num:
...
+
```
Comment (fubuloubu) [inline tests/test_parser.py]: Needs test cases where you set values in another contract

File: tests/test_parser.py [hunk 4]
```diff
@@ -1390,6 +1390,115 @@ def returnten() -> num:
...
+
```
Comment (fubuloubu) [inline tests/test_parser.py]: Needs test cases where the reference is updated/changed. I actually think the way you have it now, any two contracts with the right type signature could be passed interchangably in no problem.

File: tests/test_parser.py (hunk 1 above)
Comment (JackieXu) [reply tests/test_parser.py]: It appears some parser functionality has been broken, because the expected value (from looking at the above contract) should be `5738`. It seems that getting `len()` of `bytes` is broken.

I tried the following using the code from the PR, and it resulted in `0 == 3`:

```
def test_len_bytes():
    test_contract = """
x: num

def __init__(y: bytes <= 3):
    self.x = len(y)
    
def get_len() -> num:
    return self.x
    """
    c = get_contract(test_contract, args=['bob'])
    assert c.get_len() == 3
```

File: viper/parser.py [hunk 5]
```diff
@@ -244,12 +288,12 @@ def parse_tree_to_lll(code, origcode):
 
 # Checks that an input matches its type
 def make_clamper(datapos, mempos, typ, is_init=False):
-    if not is_init:
```
Comment (JackieXu) [inline viper/parser.py]: What is the reason for removing the `codeload`? This appears to break functionality. If I add the two if-branches back in again, the tests pass.

File: tests/test_parser.py (hunk 1 above)
Comment (DavidKnott) [reply tests/test_parser.py]: Thanks for the help finding/clarifying this mistake, I fixed it.  What went wrong before was that I did the math myself in a python console and must've made a typo.

**Comment by DavidKnott** [comment]
@fubuloubu I broke the tests out into a separate file and added the additional tests you suggested.  I'd like to implement the contract data type functionality we talked about in another PR to make it easier to review.

@JackieXu I fixed the test I broke in `test_parser.py`, it was a silly mistake.

If either of you see any other tests that should be made or better ways to do things please let me know.

File: viper/parser.py [hunk 6]
```diff
@@ -162,45 +162,68 @@ def mk_getter(varname, typ):
...
+        if isinstance(item, ast.FunctionDef):
+            _defs.append(item)
+        else:
+            raise StructureException("Invalid contract reference", item)
+    return _defs
+
+
+def add_globals_and_events(_defs, _events, _getters, _globals, item):
+    if isinstance(item.annotation, ast.Call) and item.annotation.func.id == "__log__":
+                if _globals or len(_defs):
```
Comment (JackieXu) [inline viper/parser.py]: This appears to have more indentation that it should.

File: viper/parser.py [hunk 7]
```diff
@@ -280,31 +303,48 @@ def mk_full_signature(code):
...
             sigs[event.target.id] = EventSignature.from_declaration(event)
+    for _contractname in _contracts.keys():
+        _c_defs = _contracts[_contractname]
+        _defnames = [_def.name for _def in _c_defs]
+        contract = {}
+        if len(set(_defnames)) < len(_c_defs):
+            raise VariableDeclarationException("Duplicate function name: %s" % [name for name in _defnames if _defnames.count(name) > 1][0])
+        c_otherfuncs = [_def for _def in _c_defs if not is_initializer(_def)]
+        if c_otherfuncs:
+            add_gas = initializer_lll.gas
```
Comment (JackieXu) [inline viper/parser.py]: What is `add_gas` used for?

File: viper/parser.py [hunk 8]
```diff
@@ -280,31 +303,48 @@ def mk_full_signature(code):
...
     initfunc = [_def for _def in _defs if is_initializer(_def)]
     # Regular functions
     otherfuncs = [_def for _def in _defs if not is_initializer(_def)]
-    # Create the main statement
     sigs = {}
-    o = ['seq']
     if _events:
         for event in _events:
             sigs[event.target.id] = EventSignature.from_declaration(event)
+    for _contractname in _contracts.keys():
```
Comment (JackieXu) [inline viper/parser.py]: I don't think there's a need to get a copy of the keys here, right?

This should suffice:

```python
for _contractname in _contracts:
```

File: viper/parser.py (hunk 7 above)
Comment (DavidKnott) [reply viper/parser.py]: Nothing, I should take it out.
